storage/logservice/match/label_test.go

```diff
@@ -0,0 +1,174 @@
+package match
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/kuoss/lethe/letheql/model"
+	"github.com/kuoss/lethe/letheql/parser"
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+func FuzzNodeLabelMatchEqual(f *testing.F) {
+	nodeSel := &model.LogSelector{
+		Name: "node",
+		LabelMatchers: []*labels.Matcher{
+			&labels.Matcher{
+				Type:  labels.MatchEqual,
+				Name:  "process",
+				Value: "label_value",
+			},
+		},
+		LineMatchers: []*model.LineMatcher{
+			&model.LineMatcher{
+				Op:    parser.PIPE_REGEX,
+				Value: "line_value",
+			},
+		},
+		TimeRange: model.TimeRange{},
+	}
+
+	funcs, err := getLabelMatchFuncs(nodeSel)
+	if err != nil {
+		f.Fail()
+	}
+
+	verfityFunc := funcs[0]
+
+	testcases := []string{"label_value", "non_label_value"}
+	for _, tc := range testcases {
+		f.Add(tc)
+	}
+	f.Fuzz(func(t *testing.T, label string) {
+		isMatch := verfityFunc(label)
+		/*
+			if label == "label_value" && isMatch == true {
+				// as we expected
+				// comment this branch for lint
+			}
+		*/
+		if label != "label_value" && isMatch == true {
+			t.Errorf("%q label is matched with ", label)
+		}
+	})
+}
+
+func FuzzNodeLabelMatchNotEqual(f *testing.F) {
+	nodeSel := &model.LogSelector{
+		Name: "node",
+		LabelMatchers: []*labels.Matcher{
+			&labels.Matcher{
+				Type:  labels.MatchNotEqual,
+				Name:  "process",
+				Value: "label_value",
+			},
+		},
+		LineMatchers: []*model.LineMatcher{
+			&model.LineMatcher{
+				Op:    parser.PIPE_REGEX,
+				Value: "line_value",
+			},
+		},
+		TimeRange: model.TimeRange{},
+	}
+
+	funcs, err := getLabelMatchFuncs(nodeSel)
+	if err != nil {
+		f.Fail()
+	}
+
+	verfityFunc := funcs[0]
+
+	testcases := []string{"label_value", "non_label_value"}
+	for _, tc := range testcases {
+		f.Add(tc)
+	}
+	f.Fuzz(func(t *testing.T, label string) {
+		isMatch := verfityFunc(label)
+		/*
+			if label != "label_value" && isMatch == true {
+				// as we expected
+				// comment this branch for lint
+			}
+		*/
+		if label != "label_value" && isMatch == false {
+			t.Errorf("label is not matched with %q", label)
+		}
+	})
+}
+
+func FuzzNodeLabelMatchRegexOnlyContains(f *testing.F) {
+	nodeSel := &model.LogSelector{
+		Name: "node",
+		LabelMatchers: []*labels.Matcher{
+			&labels.Matcher{
+				Type:  labels.MatchRegexp,
+				Name:  "process",
+				Value: ".+bar",
+			},
+		},
+		LineMatchers: []*model.LineMatcher{
+			&model.LineMatcher{
+				Op:    parser.PIPE_REGEX,
+				Value: "line_value",
+			},
+		},
+		TimeRange: model.TimeRange{},
+	}
+
+	funcs, err := getLabelMatchFuncs(nodeSel)
+	if err != nil {
+		f.Fail()
+	}
+
+	verfityFunc := funcs[0]
+
+	testcases := []string{"foobar", "foobaar", "regardless"}
+	for _, tc := range testcases {
+		f.Add(tc)
+	}
+	f.Fuzz(func(t *testing.T, label string) {
+		isMatch := verfityFunc(label)
+		if !strings.Contains(label, "bar") && isMatch == true {
+			t.Errorf("label is matched with %q unexpectedly", label)
+		}
+	})
```

Comment:
"bar"가 없는데 매치되면 에러로 보는 거네요.
현재도 괜찮은데 보입니다만,
아래와 같이 수정해보는 건 어떨까요?

```suggestion
	re := regexp.MustCompile(".+bar")
	f.Fuzz(func(t *testing.T, label string) {
		isMatch := verfityFunc(label)
		if re.FindString(label) == "" {
			assert.False(t, isMatch)
		} else {
			assert.True(t, isMatch)
		}
	})
```

Reply:
원래 이 fuzz 테스트 작성의 의도가 우리가 지정한 `labels.MatchType`에 해당하는 `matchFunc`을 리턴하는가였는데요.
assert하는 곳에서 리턴받아야하는 함수와 동일한 로직을 작성하고(`re.MatchString`), 리턴받은 함수와 같은 결과를 내는지 fuzzing하는 것도 의미가 있어보이네요. 결국 그게 내가 원하는 함수를 얻었다는 의미가 될 수 있을 것 같아요.